internal/kubernetes/dynamic_client.go

```diff
@@ -0,0 +1,167 @@
+
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+
+	"github.com/numaproj/numaplane/internal/util/logger"
+)
+
+//todo: add unit tests
+
+func parseApiVersion(apiVersion string) (string, string, error) {
+	// should be separated by slash
+	index := strings.Index(apiVersion, "/")
+	if index == -1 {
+		// if there's no slash, it's just the version, and the group should be "core"
+		return "core", apiVersion, nil
+	} else if index == len(apiVersion)-1 {
+		return "", "", fmt.Errorf("apiVersion incorrectly formatted: unexpected slash at end: %q", apiVersion)
+	}
+	return apiVersion[0:index], apiVersion[index+1:], nil
+}
+
+type GenericObject struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   runtime.RawExtension `json:"spec"`
+	Status runtime.RawExtension `json:"status,omitempty"`
+}
+
+type GenericStatus struct {
+	Phase      string             `json:"phase,omitempty"`
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+func ParseStatus(obj *GenericObject) (GenericStatus, error) {
+	var status GenericStatus
+	err := json.Unmarshal(obj.Status.Raw, &status)
+	if err != nil {
+		return GenericStatus{}, err
+	}
+	return status, nil
+}
+
+// ApplyCRSpec either creates or updates an object identified by the RawExtension, using the new definition,
+// first checking to see if there's a difference in Spec before applying
+func ApplyCRSpec(ctx context.Context, restConfig *rest.Config, object *GenericObject, pluralName string) error {
+	numaLogger := logger.FromContext(ctx)
+
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create dynamic client: %v", err)
+	}
+
+	group, version, err := parseApiVersion(object.APIVersion)
+	if err != nil {
+		return err
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: pluralName,
+	}
+
+	// Get the object to see if it exists
+	resource, err := client.Resource(gvr).Namespace(object.Namespace).Get(ctx, object.Name, v1.GetOptions{})
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// create object as it doesn't exist
+			numaLogger.Debugf("didn't find resource %s/%s, will create", object.Namespace, object.Name)
+
+			unstruct, err := objectToUnstructured(object)
+			if err != nil {
+				return err
+			}
+
+			_, err = client.Resource(gvr).Namespace(object.Namespace).Create(ctx, unstruct, v1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to create Resource %s/%s, err=%v", object.Namespace, object.Name, err)
+			}
+			numaLogger.Debugf("successfully created resource %s/%s", object.Namespace, object.Name)
+		} else {
+			return fmt.Errorf("error attempting to Get resources; GVR=%+v", gvr)
+		}
+
+	} else {
+		numaLogger.Debugf("found existing Resource definition for %s/%s: %+v", object.Namespace, object.Name, resource)
+		// todo:
+		//   If the existing annotation matches the new hash, then nothing to do: log and return
```

Comment:
```suggestion
		
```

Reply:
sorry, I originally did intend for that logic to live here, but I think we will have the annotation on the Rollout CR instead

Reply:
I think this may have accidentally been resolved? (we should remove the comments)

Reply:
ah I thought you want me to update the comment. will add a PR for it.